File: src/main/java/MergeTwoSortedLists/BIGBOSS/MergeTwoSortedLists.cpp
//
// Created by BIGBOSS on 2019-06-27.
//

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
    public:

    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2)
    {
        if (!l1 && l2) return l2;
        if (l1 && !l2) return l1;
        if (!l1 && !l2) return nullptr;

        ListNode* result = l1->val < l2->val ? l1 : l2;
        ListNode* cur = result;

        if (l1 == cur)
            l1 = l1->next;
        else
            l2 = l2->next;

        while (l1 && l2)
        {
            if (l1->val < l2->val)
            {
                cur->next = l1;
                l1 = l1->next;
            }
            else
            {
                cur->next = l2;
                l2 = l2->next;
            }

            cur = cur->next;
        }

        if (l1)
            cur->next = l1;

        if (l2)
            cur->next = l2;

        return result;
    }
};

File: src/main/java/MergeTwoSortedLists/panxin/MergeTwoSortedLists.java
package MergeTwoSortedLists.panxin;


public class MergeTwoSortedLists {

    public static void main(String[] args) {
        //1->2->4, 1->3->4
        ListNode l1 = new ListNode(2);
        l1.next = new ListNode(3);
        l1.next.next = new ListNode(9);
        l1.next.next.next = new ListNode(13);
        l1.next.next.next.next = new ListNode(14);
        l1.next.next.next.next.next = new ListNode(15);
        l1.next.next.next.next.next.next = new ListNode(16);

        ListNode l2 = new ListNode(2);
        l2.next = new ListNode(3);
        l2.next.next = new ListNode(4);

        ListNode listNode = new Solution().mergeTwoLists(l1, l2);
        System.out.println(listNode);
        print(listNode);
    }

    static void print(ListNode listNode) {
        if (listNode.next != null) {
            System.out.print(listNode.val + ", ");
            print(listNode.next);
        } else {
            System.out.print(listNode.val);
        }

    }

}
